syntax = "proto3";

package wxbox;

import "google/protobuf/descriptor.proto";

extend google.protobuf.FileOptions {
    string WxBoxServerURI = 10000;
}

// default WxBoxServer URI
option (WxBoxServerURI) = "localhost:52333";

service WxBox {
    rpc Ping(WxBoxHelloRequest) returns (WxBoxHelloResponse) {}
    rpc Communication(stream WxBotControlPacket) returns (stream WxBoxControlPacket) {}
}

enum ControlPacketType {
    HANDSHAKE_REQUEST = 0;
    HANDSHAKE_RESPONSE = 1;
    PROFILE_REQUEST = 2;
    PROFILE_RESPONSE = 3;
    UNINJECT_WXBOT_REQUEST = 4;
    LOGOUT_WECHAT_REQUEST = 5;
    ALL_CONTACT_REQUEST = 6;
    ALL_CONTACT_RESPONSE = 7;
    INJECT_ARGS_REQUEST = 8;
    INJECT_ARGS_RESPONSE = 9;
    EXECUTE_PLUGIN_SCRIPT_REQUEST = 10;
    EXECUTE_PLUGIN_SCRIPT_RESPONSE = 11;
    LOG_REQUEST = 12;
    CHANGE_CONFIG_REQUEST = 13;
    CLEAR_COMMAND_LOG_REQUEST = 14;
}

enum WxBotLogLevel {
    Information = 0;
    Warning = 1;
    Error = 2;
}

message WxBoxHelloRequest {
    uint64 pid = 1;
}

message WxBoxHelloResponse {
    bool pong = 1;
}

message HandshakeRequest {
    
}

message HandshakeResponse {
    bool success = 1;
}

message ProfileResponse {
    bool logined = 1;
    string nickname = 2;
    string wxnumber = 3;
    string wxid = 4;
}

message ContactProfile {
    bool chatroom = 1;
    string nickname = 2;
    string wxnumber = 3;
    string wxid = 4;
    string remark = 5;
}

message AllContactResponse {
    repeated ContactProfile contacts = 1;
}

message InjectArgsResponse {
    bytes args = 1;
}

message ExecutePluginScriptRequest {
    string statement = 1;
}

message ExecutePluginScriptResponse {
    string result = 1;
}

message LogRequest {
    WxBotLogLevel level = 1;
    string msg = 2;
}

message ChangeConfigRequest {
    bool avoidRevokeMessage = 1;
}

message WxBotControlPacket {
    ControlPacketType type = 1;
    uint64 pid = 2;
    oneof packet {
        HandshakeRequest handshakeRequest = 10;
        ProfileResponse profileResponse = 11;
        AllContactResponse allContactResponse = 12;
        InjectArgsResponse injectArgsResponse = 13;
        ExecutePluginScriptResponse executePluginScriptResponse = 14;
        LogRequest logRequest = 15;
    }
}

message WxBoxControlPacket {
    ControlPacketType type = 1;
    oneof packet {
        HandshakeResponse handshakeResponse = 10;
        ExecutePluginScriptRequest executePluginScriptRequest = 11;
        ChangeConfigRequest changeConfigRequest = 12;
    }
}